lms/services/canvas.py

```diff
@@ -1,41 +1,198 @@
-from lms.services.exceptions import CanvasFileNotFoundInCourse
+from functools import cached_property, lru_cache
+from logging import getLogger
+
+from sqlalchemy import and_
+
+from lms.models import File
+from lms.services.exceptions import CanvasAPIPermissionError, CanvasFileNotFoundInCourse
+
+LOG = getLogger(__name__)
 
 
 class CanvasService:
     """A high level Canvas service."""
 
     api = None
 
-    def __init__(self, canvas_api):
+    def __init__(
+        self, canvas_api, application_instance_service, assignment_service, db_session
+    ):
         self.api = canvas_api
 
-    def public_url_for_file(self, file_id, course_id, check_in_course=False):
+        self._application_instance_service = application_instance_service
+        self._assignment_service = assignment_service
+        self._db_session = db_session
+
+    def public_url_for_file(
+        self, file_id, course_id, resource_link_id, check_in_course=False
+    ):
         """
         Get a public URL for a Canvas file.
 
-        This will also attempt to check if the file is in the course to detect
-        course copy situations and fix it if we can.
+        This will also attempt to use any file with the same size and name if
+        one can be found.
 
         :param file_id: The file to look up
         :param course_id: The course the file should be in
-        :param check_in_course: Raise CanvasFileNotFoundInCourse if file_id isn't in
-            course_id
+        :param resource_link_id: The id of the assignment
+        :param check_in_course: Attempt to map detect and map the file if it's
+            not in the course before attempting to retrieve it. This is more
+            thorough, but causes an extra API request
         :return: A URL suitable for public presentation of the file
-        :raise  CanvasFileNotFoundInCourse: if check_in_course=True and file_id isn't in course_id
-            be in the course but isn't.
+        :raise CanvasFileNotFoundInCourse: if the file is not accessible and
+            cannot be mapped to a file in the course
+        :raise CanvasAPIResourceNotFound: if the file is not valid at all, and
+            cannot be mapped to a file in the course
+        :raise CanvasAPIPermissionError: If we cannot retrieve the file for
+            any other reason
         """
+        # Check to see if we've mapped this file id to something else before
+        file_mapping = self._file_mapping(resource_link_id)
+        mapped_id = file_mapping.get(file_id, default=file_id)
+
+        # If we are told to check up front, try and fix the file if it's not in
+        # the current course (as far as we can tell from this users point of
+        # view)
+        already_mapped = False
+        if check_in_course and not self._is_file_in_course(mapped_id, course_id):
+            mapped_id = self._match_file_in_course(file_id, course_id, file_mapping)
```

Comment:
Much of the confusing complexity of this process is the fact that we perform the mapping in two locations. Here if `check_in_course` is enabled and the file isn't in the course, and later on if we didn't do this.

This means a teacher could get either test, depending on whether the file is in the course or not. A file in the course that somehow (how?) can't be accessed by the teacher might trigger the next search. Although I think this is actually impossible, as if the file is "in the course" from the point of view of the teacher, they can see it. Nevertheless, I don't want to rely on Canvas being sane.